contracts/MarketOracle.sol

```diff
@@ -14,104 +14,128 @@ interface IOracle {
 /**
  * @title Market Oracle
  *
- * @dev Provides the aggregated exchange rate data onchain using data from a whitelisted
- *      set of market source contracts.
- *      Exchange rate is the TOKEN:TARGET rate.
+ * @dev Provides the aggregated rate data onchain using data from a whitelisted
+ *      set of providers.
  */
 contract MarketOracle is Ownable, IOracle {
     using SafeMath for uint256;
 
-    // Whitelist of sources
-    MarketSource[] public whitelist;
+    struct Report {
+        address provider;
+        uint256 timestamp;
+        uint256 payload;
+    }
 
-    event LogSourceAdded(MarketSource source);
-    event LogSourceRemoved(MarketSource source);
-    event LogSourceExpired(MarketSource source);
+    mapping (address => uint256) internal whitelistMap;
+    Report[] public whitelistReports;
+
+    event LogSourceAdded(address source);
+    event LogSourceRemoved(address source);
+    event LogSourceExpired(address source);
+
+
+    // The number of seconds after which the report must be deemed expired.
+    uint256 public reportExpirationTimeSec = 6*60*60;
+
+    uint256 public minimumProviders = 1;
```

Comment:
Necessary?

Reply:
I am assuming you are referring to having minimumProviders, not the value 1.
I think data integrity wise it makes sense to only accept the value when multiple providers are contributing to the calculation. Initially we should set minimumProviders to the total number of feeds we have so (they all have to not be stale for rebase to go through) until we  have more than 3 feeds then we can have minimimProviders < whitelistSize

Reply:
All that has been considered. What is the alternative you are in favor for? (B) No parameter and only accept the value if all the whitelist providers are all valid  or (C) consider the data valid with at least 1 provider valid. I think the later is very vulnerable.The former is the most operational maintenance, as whitelist removal will be the only possible mitigation for stale provider, hence more need for governance intervention. Current parameter approach is (A). I can stand behind A or C. leaning towards A or going with C with explicit intention that when we have more providers we upgrade to A or another approach to have more tolerance than C

Reply:
I don't have a strong opinion on having vs not having. So okay having it as it gives us a bit more control.  👍